### ISSUE TITLE
10$: error if constructor has TypeScript 'this' parameter

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#853]
We should diagnose the error in the following lines:
```typescript
type F = new (this: string) => string;
class C {
  constructor(this: string) {}
}
```

**Comment by Guilherme-Vasconcelos** [comment]
Hey, I'd like to give this issue a try, can you assign it to me?

I will decline the payment this time, though

**Comment by Guilherme-Vasconcelos** [comment]
Do you think it would be a nice addition if, when `this` is misplaced in a constructor, the new "`this` is not allowed in constructor" error gets emitted instead of E0303? Or should that change be a different issue/PR?

```typescript
// Right now these emit `E0303`
type T = new (f: string, this: string) => string;
class C {
  constructor(f: string, this: string) {}
}
```

**Comment by strager** [comment]
> Do you think it would be a nice addition if, when `this` is misplaced in a constructor, the new "`this` is not allowed in constructor" error gets emitted instead of E0303?

Yes, one diagnostic is better than two. We already do this; in non-TypeScript mode, we emit E0305 ("'this' parameters are not allowed in JavaScript") only.

> And if so, should that change be a different issue/PR?

I estimate that it should be easy to put it in the same PR. But if it turns out to be complicated, a separate PR is fine.

**Comment by Guilherme-Vasconcelos** [comment]
@strager sorry about that, but I will have to unassign myself from this issue

I've been focusing on some other stuff during my free time and haven't done much programming in a while, so I won't be able to finish this

I'll keep my fork available in case the person who picks this wants to copy the test cases
